### PR TITLE
fix: changing default numWorkers, making it customizable and refactor…

### PR DESCRIPTION
**Pull request by aokirisaki**
…ing scheduler

**What type of PR is this?**

**Which issue does this PR fix**:
#1486 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/extensions/extension_search_test.go [hunk 1]
```diff
@@ -56,6 +57,9 @@ func TestTrivyDBGenerator(t *testing.T) {
 		sch.SubmitGenerator(generator, 12000*time.Millisecond, scheduler.HighPriority)
 
 		ctx, cancel := context.WithCancel(context.Background())
+		cfg := config.New()
```
Comment (andaaron) [inline pkg/extensions/extension_search_test.go:60]: Is this config object used?

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -68,13 +70,15 @@ type Scheduler struct {
...
+	NumWorkers        int
 }
 
-func NewScheduler(logC log.Logger) *Scheduler {
+func NewScheduler(cfg *config.Config, logC log.Logger) *Scheduler {
 	chLow := make(chan Task, rateLimiterScheduler)
 	chMedium := make(chan Task, rateLimiterScheduler)
 	chHigh := make(chan Task, rateLimiterScheduler)
 	generatorPQ := make(generatorsPriorityQueue, 0)
+	NumWorkers := getNumWorkers(cfg)
```
Comment (andaaron) [inline pkg/scheduler/scheduler.go:81]:
```suggestion
	numWorkers := getNumWorkers(cfg)
```

File: pkg/extensions/extension_search_test.go [hunk 3]
```diff
@@ -31,7 +32,7 @@ func TestTrivyDBGenerator(t *testing.T) {
 		defer os.Remove(logFile.Name()) // clean up
 
 		logger := log.NewLogger("debug", logFile.Name())
```
Comment (andaaron) [inline pkg/extensions/extension_search_test.go:34]: Please use a logger which sends messages to both stdout and the logfile?
````
writers := io.MultiWriter(os.Stdout, logFile)
logger.Output(writers)
```
There are some tests using similar logic in search_test.go.
It's hard to figure out why the test is failing if the logs are not shown in the console.

**Review by andaaron — COMMENTED** [review]
There's a test case failure and it's hard to troubleshoot without seeing the logs.

**Review by andaaron — APPROVED** [review]
LGTM

File: pkg/extensions/extension_search_test.go [hunk 4]
```diff
@@ -53,16 +60,17 @@ func TestTrivyDBGenerator(t *testing.T) {
 		cveInfo := cveinfo.NewCVEInfo(storeController, repoDB, "ghcr.io/project-zot/trivy-db", "", logger)
 		generator := NewTrivyTaskGenerator(time.Minute, cveInfo, logger)
 
-		sch.SubmitGenerator(generator, 12000*time.Millisecond, scheduler.HighPriority)
+		sch.SubmitGenerator(generator, 1200*time.Millisecond, scheduler.HighPriority)
```
Comment (andaaron) [inline pkg/extensions/extension_search_test.go:63]: This seems too often. Let's not download the DB every 1.2 seconds? It doesn't even have enough time to complete the download before the new one starts.

**Review by andaaron — DISMISSED** [review]

**Review by andaaron — APPROVED** [review]

**Review by rchincha — COMMENTED** [review]
This looks good. However, is there is a task scheduler stress test we can add?

**Comment by andaaron** [comment]
> This looks good. However, is there is a task scheduler stress test we can add?

Lets track this separately. The initial task is complete.

**Pull request merged**